### ISSUE TITLE
New Operator: ParseDateTime

### ISSUE THREAD

**Issue by cbourjau** [onnx/onnx#5409]
# ParseDateTime

### Describe the operator
An operator which parses a date time string into a Unix time stamp.

The operator in this form was proposed during a roadmap 2023 [presentation](https://github.com/onnx/steering-committee/blob/main/roadmap/2023.md#18-datetime-parsing). 

The proposed operator signature is as follows:

**Inputs:**
- **x**: `tensor(string)`

**Attributes:**
- **unit:** string: second (`s`), millisecond (`ms`), microsecond (`us`) or nanosecond (`ns`); **default TBD**
- **default:** tensor, optional; A rank-0 tensor containing the default value to be used if the parsing fails. The type of the tensor is also the output type. If `default` is not given, the behavior for parsing failures is implementation-defined and the output type is `tensor(int64)`.
- **format:** string; Formatting string in ``strptime`` like syntax. The following directives are currently supported:

    **Date directives**
    - `%Y`: The full proleptic Gregorian year, zero-padded to 4 digits. (e.g. 2001)
     - `%y`: The proleptic Gregorian year modulo 100, zero-padded to 2 digits. (e.g. 01)
     - `%m`: Month number (1-12); optionally zero-padded to two digits
     - `%d`: Day of the month (1-31); optionally zero-padded to two digits

    **Time directives**
    - `%H`: Hour of the day in 24h format (0–23); optionally zero-padded to two digits
    - `%M`: Minutes of the hour (0–59); optionally zero-padded to two digits
    - `%S`: Seconds of the minute (0–59); optionally zero-padded to two digits

**Outputs:**
- **y:** `tensor(double)`, `tensor(int64)`
A tensor of the same shape as `x`. The elements are Unix time stamps in `unit` and of a type defined by the `default` attribute.

##### Use case
We use ONNX to persist feature engineering pipelines for tabular data. Datetime strings are a common occurrence in this application and are often blocking us from migrating pipelines entirely into ONNX.

##### Format string
The format-syntax of `strptime` is ubiquitous and widely used across various languages and libraries. This includes C/C++, Python and Rust (via the Chrono crate). 

##### Output type

There are several considerations to be made about the output type:
- A floating point output has the advantage of gracefully handling invalid inputs by mapping it to `NaN`. The downside is a loss in precision. 
- (Valid) Unix time stamps are often processed as integer values. Numpy, for example, uses `int64`s for this purpose and uses a "magic number" to label `NaT`-elements. However, for Numpy this is an implementation detail that is not readily exposed to the user. We don't have that level of abstraction available in ONNX. Magic number elements would be directly exposed to the user. While `NaN` values may naturally propagate to the outputs, this is not possible for `int64`.
- Using floating point time stamps is not without popular precedent. Python's `datetime.datetime.timestamp` method returns floating point values, too.
- Lastly, it may also be desired to actually abort the inference process when encountering an invalid string element. As far as I'm aware, the standard does not have any standardized error handling. The current proposal keeps the door open for such behavior via the optional `default` attribute.

##### Unit
The unit is an open question. Pandas supports seconds, milliseconds, microseconds, and nanoseconds.

### Can this operator be constructed using existing onnx operators?
No

### Is this operator used by any model currently? Which one?
We use ONNX extensively for feature engineering. Datetime parsing is a recurring pain point for us that we currently solve with pre-processing Python code. Adding this operator to the standard would allow us to ship pipelines as self-contained ONNX models. 

### Are you willing to contribute it? (Y/N)
Yes

### Prior art 
- [Attempts](https://github.com/tnakaike/onnx/blob/0728ff2681f303202ad1391fce6dc13bdc4414e3/onnx/defs/traditionalml/defs.cc#L1179-L1181) to add a date time parsing operator to ONNX have been made in the past by @tnakaike

### Future extensions
- Operators which convert time stamps to day, week, month and/or year

### Notes
- A similar (and compatible) operator exists for [Tensor Flow](https://www.tensorflow.org/addons/api_docs/python/tfa/text/parse_time)

**Edits:** 
- Updated after feedback from the Preprocessing SIG.
- Clarified definition of directives

**Comment by gramalingam** [comment]
Overall, this sounds good. The only issue that is a bit unclear (as to what is the best option) is the handling of invalid string elements. You mention Numpy returns a magic number NaT. What about the libraries in other languages mentioned (like C/C++)?

If any standard library fails by aborting, it may be hard to use it to implement this op as described, since when the user specifies a default-value, the implementation cannot abort. So, it may be useful to make a definitive call whether we want to allow for abort/exception or not.

A second thought that comes to mind is that deterministic behavior is very helpful in machine-learning and neural networks, since they can learn to handle abnormal values the right way, as long as they are represented deterministically. So, I wonder if it is better to specify explicitly the default-value used if the user doesn't specify it: eg., something like 64-bit MININT.

**Comment by gramalingam** [comment]
A few questions:

* The python doc for strptime format mentions a 1989 C standard (which seems fine), but also mentions additional directives not required by the C89 standard which may not be available on all platforms. Should we just restrict the spec to the 1989 C standard?
* What happens if the format codes supported varies across platforms (as the python doc says it will)?

This is just about the format-string, as opposed to errors in the input-data (wrt specified format).

**Comment by gramalingam** [comment]
I don't fully understand your use-case. But I notice that strptime like functions essentially return structures, keeping the year, month, day, hours, minute. second, microsecond separate. Would it be useful for us to do something similar? We can encode a structure like this as a 7-tuple or 8-tuple of ints. So, for an input of shape [N], we could return an int32-tensor of shape [N,8]. 

It seems to me that for a ML-application these different components (eg., month or day) could have vastly different role to play as features.

**Comment by gramalingam** [comment]
A couple of other comments, relating to the current proposal: strptime doesn't seem to support nanoseconds, only microseconds. Is there a real need to have nanoseconds? It's sort of relevant to the question of encoding the entire datetime as a single value (int64 or double). With microseconds, an int64 encoding seems adequate, but gets trickier with nanoseconds.

**Comment by cbourjau** [comment]
Thanks for the great feedback!

> The only issue that is a bit unclear (as to what is the best option) is the handling of invalid string elements. You mention Numpy returns a magic number NaT. What about the libraries in other languages mentioned (like C/C++)?

The most prominent example here is probably `strptime` from the c standard library. It returns 'NULL' if the parsing fails as documented [here](https://man7.org/linux/man-pages/man3/strptime.3.html#RETURN_VALUE).

> So, it may be useful to make a definitive call whether we want to allow for abort/exception or not.

I agree that a standardized section about error handling would be great. This might be something that could be part of the `backend` API, maybe? 

> Should we just restrict the spec to the 1989 C standard?

One problem with the C standard that I just noticed is that it does not support sub-second precision. In C/C++ we may use `howardhinnant-date` instead of the standard library. Finding a good standard here is more difficult that anticipated, to be honest. I will try to distill a reasonable subset tomorrow which we then add explicitly to the standard. 
  
> What happens if the format codes supported varies across platforms (as the python doc says it will)?

Allowing an explicit subset as described above would solve this issue, too.

> I don't fully understand your use-case

Our use cases often involve "traditional ML" methods and feature engineering of tabular data. For example, we may get a date string from which we need to compute an age (time delta). Sometimes, the time delta from Unix time stamps may be good enough. At other times, we may want that time delta in years or months. In that case, we may either have to resort to implementing the required computations from time stamps ourselves inside the ONNX model (for a time horizon reasonable for a given project) or propose follow-up operators for work with time stamps. 

In any case, I do believe that returning time stamps is more flexible than returning tuples. Computing something as simple as a time delta in seconds is trivial with time stamps but very involved using tuples.

**Comment by cbourjau** [comment]
I updated the proposal to use an explicit and small subset of `strptime`. This should cover many use cases while not creating an undue burden on runtimes. We may consider explicit operators for parsing RFC 2822 strings (e.g. `Wed, 18 Feb 2015 23:16:09 GMT`) or RFC 3339 strings (e.g. `1996-12-19T16:39:57-08:00`) in the future rather than exposing all necessary directives in this operator.
